packages/axios-extension/src/abap/lrep-service.ts

```diff
@@ -0,0 +1,163 @@
+import type { Service } from '../base/service-provider';
+import type { AxiosResponse } from 'axios';
+import { Axios } from 'axios';
+import { LogLevel } from '@sap-ux/logger';
+import type { Logger } from '@sap-ux/logger';
+import { readFileSync } from 'fs';
+
+/**
+ * Type representing a namespace. It is either a string or an object containing an id (variant id) and a reference (base application id).
+ */
+export type Namespace =
+    | string
+    | {
+          /**
+           * variant id
+           */
+          id: string;
+
+          /**
+           * base application id
+           */
+          reference: string;
+      };
+
+/**
+ * Required configuration to deploy an adaptation project.
+ */
+export interface AdaptationConfig {
+    /**
+     * Namespace either as string or object
+     */
+    namespace: Namespace;
+
+    /**
+     * Optional ABAP package name
+     */
+    package?: string;
+
+    /**
+     * Optional transport request
+     */
+    transport?: string;
+}
+
+/**
+ * Structure of the result message.
+ */
+export interface Message {
+    severity: 'Success' | 'Warning' | 'Error';
+    text: string;
+    details?: string[];
+    id: string;
+    variables?: string[];
+}
+
+/**
+ * Technically supported layers, however, in practice only `CUSTOMER_BASE` is used
+ */
+type Layer = 'VENDOR' | 'CUSTOMER_BASE';
+
+/**
+ * Returns the namespace as string.
+ *
+ * @param namespace either as string or as object
+ * @returns serialized namespace
+ */
+function getNamespaceAsString(namespace: Namespace): string {
+    return namespace['id'] ? `apps/${namespace['reference']}/appVariants/${namespace['id']}/` : (namespace as string);
+}
+
+/**
+ * A class respresenting the design time adaptation service allowing to deploy adaptation projects to an ABAP system.
+ */
+export abstract class LayeredRepositoryService extends Axios implements Service {
+    public static readonly PATH = '/sap/bc/lrep';
+
+    public log: Logger;
+
+    /**
+     * Check whether a variant with the given namespace already exists.
+     *
+     * @param namespace either as string or as object
+     * @returns the Axios response object for futher processing
+     */
+    public async check(namespace: Namespace): Promise<AxiosResponse> {
+        return this.get('/dta_folder/', {
+            params: {
+                name: getNamespaceAsString(namespace),
+                layer: 'CUSTOMER_BASE' as Layer
+            }
+        });
+    }
+
+    /**
+     * Fetch a csrf token required for deployment. The token will be returned as header 'x-csrf-token'.
+     *
+     * @returns the Axios response object for futher processing
+     */
+    public async getCsrfToken(): Promise<AxiosResponse> {
```

Comment:
Does this method need to be public?

Reply:
Does it need to be, no, but the guiding principle for the axios-extension is to offer convenience but still keep it flexible, therefore, I would like to keep it public

Reply:
This service does the CSRF token handling differently to the other services/providers:https://github.com/SAP/open-ux-tools/blob/d9d13e7127d5cea330574622f5d9e1e0dd7bb7fe/packages/axios-extension/src/auth/connection.ts#L137-L140

Is it possible to extend the current implementation?

The intention of asking if we want to expose it is because it makes sense to keep the API surface area as small as possible (specifically I had [Hyrum's law](https://www.hyrumslaw.com/) in mind)

Reply:
You are absolutely correct, the service specific CSRF handling was unnecessary. I have completely removed it. The CSRF token will be fetched with the first `GET` request which we anyhow always send to check if this is a new or existing adaptation project.
When removing it, I also noticed a slightly incorrect implementation in `Connection.ts`. Our code threw an exception for every http code `>=400`, however, that is wrong because the connection can be successful even if a `404` or a `500` is then returned, therefore, I have limited the check to `=401`.

Change is in https://github.com/SAP/open-ux-tools/pull/469/commits/b8ee2925a0889471a628b6eb9a0444917d1e9e83